Kevin merge upstream

knisbet opened the pull request:
Update our fork with changes from the upstream flannel project.

Fixes gravitational/telekube#3527

MohdAhmad committed:
Vendoring in bronze1man goStrongswanVici repo.

This commit adds the IPSEC backend.

IPSEC Mode: Tunnel
This means that the entire IP packet is encrypted. We encapsulate using
the host IP addresses therefore we can have a separate IPSEC backend
instead of integrating it with existing backends. Running in tunnel mode
has the added advantage of not leaking the container IPs.

Authentication Mechanism: Pre-shared key (PSK)
For authentication, we use a PSK, generated by flannel and distributed
through etcd which in turn uses client-side certificates for
authentication.

Internet key exchange (IKE): Charon
We use charon daemon from the strongSwan project to negotiate Security
Associations (SAs) between the endpoints. SAs define the security
attributes, used to enable secure communication between endpoints [1].
IKE has two phases, the first establishes a secure channel using the
Diffie-Hellman key exchange and the second phase uses this secure
channel to negotiate the security attributes like, for example the keys
used to encrypt traffic. We use, 4096 bit Diffie-Hellman, AES-256-CBC
for encryption and SHA-256 for integrity. Note IKEv1 is used and phase 1
is run in Main mode.

Configuration
The only configuration parameter required is whether UDP Encapsulation
should be enabled.

Commit:
Migrating to use glide vendoring

Update to use new Makefile based build

Fixup k8s mgr for updated interface

tomdee committed:
backend/vxlan: simplify vxlan processing

mkutsevol committed:
Apply changes to etcdv2 code after package move

Packaging and support for ikev2

Vendoring goStrongswanVici

Make ESP proposals configurable & default to Suite-B-GCM-128

goStrongswanVici dependency update

Removed CreateBackendData/GetBackendData

 * As flannel doesn't support multiple networks, this is not needed any
more

Splitting launch and management of charon

 * Transitional, (WIP)
 * Support for bundled/remote charon.
 * Cleanup after removal of CreateBackendData/GetBackendData

Some packaging

 * It builds amd64.
 * Still much to do.

RegisterNetwork can correctly shutdown spawned processes

 * Added sync.WaitGroup, so spawned processes can correctly shutdown.
 * Bundled charon daemon correctly shuts down.

ChenLingPeng committed:
Add new ipip backend

Signed-off-by: forrestchen <forrestchen@tencent.com>
Signed-off-by: Chun Chen <ramichen@tencent.com>

IPIP backend update

- Rename ipip device to flannel.ipip
- Add tests for ipip backend
- Add mtu config to ipip backend
- Fix invalid use of EEXIST
- Remove the equal check of `IfaceAddr` and `ExtAddr`
- Set MTU to 20 less than the MTU of the selected iface (specified with the --iface option)

Update according to the review

- Recreate flannel.ipip device if local address changes
- Add `EnsureV4AddressOnLink`
- Update ipip document to describe that the tunl0 device will be created but that's not a bug

Merge hostgw network with ipip network

andyxning committed:
add healthz

tomdee committed:
Merge pull request #722 from andyxning/add_healthz

add healthz

tomdee committed:
Makefile: Smaller flanneld binary

Use Go 1.8.x
Strip the binary

Improvement is ~50MB -> ~30MB

mgleung committed:
Add iptables rule to prevent masquerade on some external traffic

This fixes a bug that was preventing OnlyLocal annotation in Kubernetes
from working correctly. Prevents masquerade on external traffic to
containers/pods in the subnet.

Fixes #734

oilbeater committed:
fix typo

gunjan5 committed:
Merge pull request #748 from oilbeater/oilbeater-patch-1

fix typo

tomdee committed:
Merge pull request #744 from tomdee/smaller-binaries

Makefile: Smaller flanneld binary

mgleung committed:
Added compatibility check to compare GBP for VXLAN

vxlanLinkIncompat function did not check GBP so it was possible that
changes in the config would not be refreshed.

Fixes #562

tomdee committed:
Merge pull request #750 from ironlung975/vxlan-check-gbp

Added compatibility check to compare GBP for VXLAN

tomdee committed:
Merge pull request #745 from ironlung975/onlylocal-annotation

Add iptables rule to prevent masquerade on some external traffic

mgleung committed:
flannel reads from created subnet.env file on startup

Added feature to allow flannel to restart in case of etcd failures and
still keep the same subnet address for the hosts.

Fixes #610 #29

mgleung committed:
Added iface-regex option

Allows for more flexibility when selecting the network interface that
flannel should be using.

Addresses #620

tomdee committed:
Merge pull request #754 from mgleung/iface-regex

Added iface-regex option

tomdee committed:
Merge pull request #752 from mgleung/prev-lease

flannel reads from created subnet.env file on startup

mgleung committed:
Added ability to specify multiple ifaces and iface regexes

Allows for even more flexibility when selecting the network interface
that flannel should be using

Addresses #620

mgleung committed:
Fix a bug with the iface-regex that always returned an error

tomdee committed:
Merge pull request #756 from mgleung/fix-iface-regex

Fix a bug with the iface-regex that always returned an error

tomdee committed:
Allow kube subnet manager to run outside of kubernetes

This is really useful for dev and test work

This PRs also fixes some problems in the vendoring and adds a target to
the Makefile to make it easy to run.

tomdee committed:
Docs: Add kubernetes and troubleshooting info

mgleung committed:
Fix a bug where previously leased subnets would not update etcd leases

tomdee committed:
main.go: Fix logging options

Only display the logging options that actually work

tomdee committed:
Merge pull request #762 from tomdee/fix-log-options

main.go: Fix logging options

tomdee committed:
Merge pull request #759 from tomdee/support-k8s-out-of-cluster

Allow kube subnet manager to run outside of kubernetes

tomdee committed:
Merge pull request #761 from mgleung/public-ip-change

Fix a bug where previously leased subnets would not update etcd leases

tomdee committed:
Merge pull request #757 from mgleung/iface-multi

Added ability to specify multiple ifaces and iface regexes

tomdee committed:
Merge pull request #764 from tomdee/docs-rewrite

Overhaul docs - troubleshooting, kubernetes info

tomdee committed:
Update manifest to v0.8.0

tomdee committed:
Merge pull request #768 from tomdee/v0.8.0-manifest

Update manifest to v0.8.0

tomdee committed:
Remove RevokeLease from manager interface

tomdee committed:
Remove wait() from backendmanager. Nothing calls it

tomdee committed:
main.go: Remove sleep during shutdown

evie404 committed:
use init container to install cni on flannel daemonset

tomdee committed:
Merge pull request #778 from tomdee/remove-wait

Remove wait() from backendmanager. Nothing calls it

tomdee committed:
Merge pull request #776 from tomdee/remove-revoke

Remove RevokeLease from manager interface

tomdee committed:
Merge pull request #777 from tomdee/remove-shutdown-sleep

main.go: Remove sleep during shutdown

tomdee committed:
backend/vxlan: Add support for "direct routing"

This skips vxlan encapsulation if the hosts are on the same subnet

athai committed:
Rewrite sentence and remove reference to go const

athai committed:
Add '--' to command option

oilbeater committed:
fix: add ca-certificates

Backend like aws-vpc or ali-vpc call https IaaS API will failed without ca-certificates.

Commit:
Merge pull request #789 from athai/remove-broken-reference

Remove broken reference

tomdee committed:
Merge pull request #791 from oilbeater/master

add ca-certificates

tomdee committed:
Merge pull request #785 from tomdee/vxlan

backend/vxlan: simplify vxlan processing

ruoshan committed:
Add flag validation for subnet-lease-renew-margin

when the subnet-lease-renew-margin is large than 1440(minutes), etcdv2
subnet manager would keep renew lease, causing high CPU load to flanneld
and etcd.

julia-stripe committed:
Set link index correctly on netlink messages

julia-stripe committed:
Pass old route instead of new route to RouteDel

tomdee committed:
Merge pull request #800 from ruoshan/fix/validate-subnet-lease-renew-margin

Add flag validation for subnet-lease-renew-margin

tomdee committed:
Merge pull request #803 from julia-stripe/set-link-index

Fix route deletion when replacing route in hostgw backend

tomdee committed:
Merge pull request #787 from tomdee/vxlan-direct-routing

backend/vxlan: Add support for "direct routing"

tomdee committed:
backend/hostgw: Improve robustness, add logging and comments

Ensure that the "desired" list of routes is kept up to date even if
there are errors
Add a few logging and comments

gunjan5 committed:
Merge pull request #807 from tomdee/host-gw-tidy

backend/hostgw: Improve robustness, add logging and comments

tomdee committed:
subnet/*: Remove unused reservations code

The concept of a reservation was added a long time ago but never exposed
to users. This commit removes that code.

Reservations were part of the subnet manager interface but only
implemented by the etcd backend. The Kube backend had stub code to
implement the interface but threw errors when called.

tomdee committed:
Improve functional tests

- Move to using bash_unit
- Test pings in both directions
- Test kubernetes subnet manager
- Test multiple versions of etcd and k8s during a release

julia-stripe committed:
Always ensure iptables masquerade rules are installed

mkumatag committed:
Remove lib copies

tomdee committed:
Merge pull request #811 from mkumatag/dockerfiles

Remove Library copy commands from dockefiles

gunjan5 committed:
Merge pull request #809 from tomdee/remove-reservations

subnet/*: Remove unused reservations code

tomdee committed:
Makefile: Stop pulling the unused lib from kube-cross

tomdee committed:
Merge pull request #817 from tomdee/cleanup-libs

Makefile: Stop pulling the unused lib from kube-cross

dgonzalezruiz committed:
Fix horrendous README typo

tomdee committed:
Merge pull request #822 from dgonzalezruiz/master

Fix horrendous README typo

tomdee committed:
Merge pull request #808 from julia-stripe/retry-creating-iptables-rule

Ensure that the iptables masquerade rule is always installed

tomdee committed:
Update README.md

tomdee committed:
Merge commit 'cff39eeb3d0dd3710f83b5eba7bde1afa8fa2d46' into k8s-manifests

tomdee committed:
kube-flannel.yml: Update to v0.9.0 and improve docs

- Update the manifests to the latest release
- Create a new "legacy" manifest for older versions of k8s
- Combine the RBAC info into the main manifest

tomdee committed:
Merge pull request #826 from tomdee/k8s-manifests

kube-flannel.yaml: Update manifests to latest release

sakeven committed:
Fix printf verb %s of wrong type

gunjan5 committed:
Merge pull request #827 from sakeven/patch-1

Fix printf verb %s of wrong type

sakeven committed:
Fix error message in ipmasq

Signed-off-by: sakeven <jc5930@sina.cn>

athai committed:
Make all lowercase (due to case-sensitive URLs)

Commit:
Merge pull request #839 from athai/lowercase-filename

All lowercase filename

tomdee committed:
Merge pull request #828 from sakeven/fix/error

Fix error message in ipmasq

tomdee committed:
Merge pull request #816 from tomdee/improve-tests

Improve functional tests

chenchun committed:
Remove unexpected route from route list

mkumatag committed:
Add support for multiarch e2e tests

tomdee committed:
Merge pull request #845 from mkumatag/multiarch-e2e

Add support for multiarch e2e tests

Dieken committed:
kube-flannel-cfg: enable hairpin mode

Hairpin mode is needed to allow pods to communicate with themselves via
Service IP. This change is required as the kubelet --hairpin-mode flag
is not supported when CNI is being used.

This patch is picked from https://github.com/coreos/tectonic-installer/pull/1640

alvaroaleman committed:
Add annotation to overwrite a kubelets public IP

This may be useful if a nodes public IP can not determined, e.G.
because it is behind a nat. Fixes #712

tomdee committed:
Merge pull request #849 from Dieken/master

kube-flannel-cfg: enable hairpin mode

mkumatag committed:
Fix TTY error while running in container

rakelkar committed:
network: refactor ipmasq management into it's own file

Dieken committed:
make sure flanneld got QoS class "Guaranteed" to have lower oom_score_adj

Usually flanneld consumes 5m CPU and 15Mi memory according to "kubectl top".
According to test by @tomdee at https://github.com/coreos/flannel/pull/855#issuecomment-346736586,
50Mi is enough for a 1000 node vxlan cluster.

gunjan5 committed:
Comment out the PR template message and add release note section

Commit:
README.md: fix link to kubernetes documentation

Link to kubernetes documentation was broken (wrong capitalization).

tomdee committed:
Merge pull request #850 from mkumatag/fix_docker

Fix TTY error while running e2e tests in container

tomdee committed:
Merge pull request #856 from gunjan5/pr-template

Comment out the PR template message and add release note section

tomdee committed:
Merge pull request #857 from ysim/fix-link-in-readme

README.md: fix link to kubernetes documentation

tomdee committed:
Correct broken whitespace and remove unneeded changes

tomdee committed:
travis: Only run the tests once

tomdee committed:
Makefile: Stop building ACIs

tomdee committed:
Merge pull request #859 from tomdee/remove-aci

Makefile: Stop building ACIs

tomdee committed:
Merge pull request #858 from tomdee/fix-travis

travis: Only run the tests once

tomdee committed:
Makefile: Cleanup, alpine and amd64 only UDP

Reorganize the Makefile to have different sections for build, test,
release and development.

Move to using Alpine linux as the base image. This means we no longer
need to build iptables and it makes it easier to install other packages
in future (e.g. strongswan).

To allow alpine to be used, static binaries for all platforms are
needed. It's not possible to create these for the UDP backend for
non-amd64 platforms. This is because it relies on CGO. So this PR makes
the UDP backend work on amd64 only.

mkumatag committed:
Remove leftover containers

mkutsevol committed:
Revert changes to dockerfiles.

amd64 standalone docker image with ipsec

 * Removed the build of strongswan compeletey. Even alpine linux has it
build already.
 * We package docker with strongswan as a separate image.

Switch back to setting charon executable path from cli

 * DRYer
 * Correct stop if hadn't finished init sequence.

tomdee committed:
Merge pull request #861 from mkumatag/fix_teardown

Remove leftover containers

tomdee committed:
Merge pull request #840 from alvaroaleman/fix-712

Fix #712, allow overwriting the public IP of a Kubernetes node

tomdee committed:
Merge pull request #843 from chenchun/remove

Remove unexpected route from route list

cglewis committed:
MAINTAINER is deprecated, using LABEL now

tomdee committed:
Merge pull request #863 from cglewis/master

MAINTAINER is deprecated, using LABEL now

vissible committed:
typo

gunjan5 committed:
Merge pull request #868 from vissible/patch-1

fix typo

tomdee committed:
Merge pull request #854 from rakelkar/jroggeman/ipmasq

network: move ipmasq management into platform-specific files

tomdee committed:
Merge pull request #860 from tomdee/simplify-packaging

Makefile: Cleanup, alpine and amd64 only UDP

tomdee committed:
network/iptables: Add iptables rules to FORWARD chain

To work around the Docker change from v1.13 which
changed the default FORWARD policy to DROP.

The change has bitten many many users.

The troubleshooting documentation is also updated talk about the issue.

Replaces PR #862
Fixes #834
Fixes #823
Fixes #609
Fixes #799

alvaroaleman committed:
Log kubernetes public ip override

mkumatag committed:
Fix e2e tests for non-amd64 archs

rakelkar committed:
backend: Get flannel building on windows with stubs

yuvipanda committed:
Clarify GCE route quota is only 100 by default

They'll raise it if you ask! As written, it is made to sound like a 
permanent unfixable limit.

tomdee committed:
Merge pull request #876 from mkumatag/fix_e2e

Fix e2e tests for non-amd64 archs

tomdee committed:
Merge pull request #880 from yuvipanda/patch-1

[DOC] Clarify GCE route quota is only 100 by default

tomdee committed:
Merge pull request #873 from alvaroaleman/712-logging

Log kubernetes public ip override

tomdee committed:
Merge pull request #872 from tomdee/always-ipforward

network/iptables: Add iptables rules to FORWARD chain

tomdee committed:
kube: Update manifests to v0.9.1

tomdee committed:
Merge pull request #885 from tomdee/update-manifests

kube: Update manifests to v0.9.1

tomdee committed:
Merge pull request #879 from rakelkar/jroggeman/build_windows

backend: Get flannel building on windows with stubs

Gacko committed:
Documentation/Manifests: Use list notation and command with args.

Gacko committed:
Documentation/Manifests: Enable hairpin mode in other YMLs.

tomdee committed:
Merge pull request #884 from Gacko/patch-1

Documentation/Manifests: Improve YMLs.

osoriano committed:
Add portmap plugin to CNI conf

The portmap plugin is needed for hostPorts to work

Closes: https://github.com/coreos/flannel/issues/887

tomdee committed:
Merge pull request #888 from osoriano/kube-flannel-portmap

Add portmap plugin to CNI conf

jroggeman committed:
flannel: Get flannel running on windows

tomdee committed:
backend/extension: Add Wireguard configuration

This PR adds
- the wireguard tools to the flannel container images (~0.1MB
size increase)
- an example extensions config
- a one-line change to extension.go to pass in the flannel network

The tests are commented out since they won't pass unless the wireguard
kernel module is loaded.

tomdee committed:
subnet/etcdv2/local_manager.go: Fix startup log

Print "None" instead of 0.0.0.0/0 if there's no previous subnet to use

Fixes #767

tomdee committed:
subnet/config.go: Ensure that Subnets are aligned

SubnetMin and SubnetMax must be aligned to a Subnet boundary

Fixes #430

tomdee committed:
subnet/config.go: Fix handling of subnet config

Fix all the limits, especially around minimum sizes

Fixes #696

tomdee committed:
backend/vxlan: Return correct MTU value

Ensure that the correct MTU is being used for vxlan

tomdee committed:
Merge pull request #855 from Dieken/master

make sure flanneld got QoS class "Guaranteed" to have lower oom_score_adj

tomdee committed:
Merge pull request #893 from tomdee/prev-subnet-log

subnet/etcdv2/local_manager.go: Fix startup log

gunjan5 committed:
Merge pull request #894 from tomdee/align-subnets

subnet/config.go: Ensure that Subnets are aligned

gunjan5 committed:
Merge pull request #897 from tomdee/fix-vxlan-mtu

backend/vxlan: Return correct MTU value

tomdee committed:
Code markups and minor refactorings

tomdee committed:
Merge pull request #842 from chenchun/iptun

Add new ipip backend

tomdee committed:
Merge pull request #898 from tomdee/wireguard-extension

backend/extension: Add Wireguard configuration

madhanrm committed:
Fix to build/release windows binaries from Linux

karthequian committed:
fixes typo

tomdee committed:
Merge pull request #903 from madhanrm/buildwin

Fix to build/release windows binaries from Linux

tomdee committed:
Merge pull request #907 from karthequian/patch-1

fixes typo

Commit:
Correct Spelling ClusterRoleBinding

tomdee committed:
Merge pull request #910 from abhaydiwan/patch-1

Correct Spelling ClusterRoleBinding

tomdee committed:
Fix up rebase

tomdee committed:
Lots of fixes getting it ready to merge (tests/docs)

- Add e2e tests
- Documentation
- Comments and logging tweaks
- Remove the "initial events" code, which means that stale polciy won't
  be removed.
- Various other small changes
- Get it compiling on windows

tomdee committed:
Update vendoring

tomdee committed:
Merge pull request #911 from tomdee/rakelkar-jroggeman/windows_impl

flannel: Get flannel running on windows

gunjan5 committed:
Merge pull request #896 from tomdee/fix-subnet-checking

subnet/config.go: Fix handling of subnet config

ericchiang committed:
update CoC

gunjan5 committed:
Merge pull request #912 from ericchiang/coc

automated PR: update CoC

charsyam committed:
change invalid url of kube-flannel-rbac.yml

gunjan5 committed:
Merge pull request #914 from charsyam/feature/change_url_of_rbac_yaml

Fix invalid url of kube-flannel-rbac.yml

tomdee committed:
Builds: Update go to v1.9.2

tomdee committed:
Merge pull request #915 from tomdee/update-go

Builds: Update go to v1.9.2

tomdee committed:
Update manifests to v0.10.0

tomdee committed:
Merge branch 'v0.10.0-manifests'

tomdee committed:
Merge pull request #929 from tomdee/feature/ipsec

Add IPSec backend

DiamondYuan committed:
translate alicloud-vpc-backend.md

update'

update

tomdee committed:
Merge pull request #930 from DiamondYuan/master

translate alicloud-vpc-backend.md

SleepyBrett committed:
Added new flag -iptables-resync allows you to pass an int (default 5) that is used in network/iptables SetupAndEnsureIPTables to control how often it resync's the iptables rules. We found on a larger cluster that having this value hardcoded to 5 seconds created too much contention on the iptables lock for kube-proxy to properly function

tomdee committed:
Merge pull request #935 from Nordstrom/iptresyncperiod

Added new flag -iptables-resync

dmathieu committed:
log all the available interfaces when none were found

celskeggs committed:
backend/udp: Use a /32 prefix for the flannel0 interface

This avoids the kernel's creation of broadcast routes, which prevent
communication from the host with the zeroth subnet to containers on any
other hosts.

tomdee committed:
Merge pull request #941 from sipb/fix-udp-host-access

backend/udp: Use a /32 prefix for the flannel0 interface

tomdee committed:
Merge pull request #940 from dmathieu/log-interfaces

Log all the available interfaces when none were found

alebsack committed:
Fixed typo in arguments parser

gunjan5 committed:
Merge pull request #952 from alebsack/master

Fixed typo in arguments parser

mscherer committed:
Correct spelling "reposistory"

tomdee committed:
Merge pull request #956 from mscherer/patch-1

Correct spelling "reposistory"

mousavian committed:
Fix a typo

phantooom committed:
fix typo

gunjan5 committed:
Merge pull request #975 from phantooom/master

fix typo

gunjan5 committed:
Merge pull request #970 from mousavian/master

Fix a typo

Commit:
Merge remote-tracking branch 'upstream/master' into kevin/merge-upstream

Commit:
Fix merge errors

a-palchikov reviewed (approved):
I suppose lgtm. Did you add any changes or was this all the result of the merge with upstream?

knisbet commented:
I made two changes to support the merge:

1. Added a license to kube.go, so that the project would pass upstream unit tests
2. Changed some of our added flags to match what flannel was doing. These changes can be seen on this commit where I fix some of the merge errors from the files I had to manually merge: https://github.com/gravitational/flannel/pull/3/commits/c42993ac01cf116ff6c32292800af12bdfa53301
